core/shared/src/main/scala/cats/parse/Parser.scala

```diff
@@ -1731,6 +1764,46 @@ object Parser {
     */
   def end: Parser0[Unit] = Impl.EndParser
 
+  /** return the position the parser is at
+    */
+  def careting: Parser0[Caret] = Impl.WithCaret
+
+  /** return the result of the parser together
+    * with the position of the starting of the consumption
+    */
+  def careted0[A](pa: Parser0[A]): Parser0[(A, Caret)] =
```

Comment:
My idea was that `pa` was primary and the caret was secondary so the result of `pa` comes first, but yes that would cause confusion about where is the caret exactly. If you prefer putting the caret at the first element I'll change it.

I think `pa.careted` is more readable than `(caret ~ pa)` in the sense that it associates tighter in sight which indicates it is somehow more of a single parsing unit. Indeed we can leave this out and let the user make the choice whether to write this helper instead since it is really short.